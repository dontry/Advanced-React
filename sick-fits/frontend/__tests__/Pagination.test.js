import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import Pagination, { PAGINATION_QUERY } from "../components/Pagination";
import Router from "next/router";

Router.router = {
  push() {},
  prefetch() {}
};

const makeMocksWithLength = length => [
  {
    request: {
      query: PAGINATION_QUERY
    },
    result: {
      data: {
        itemsConnection: {
          __typename: "aggregate",
          aggregate: {
            count: length,
            __typename: "count"
          }
        }
      }
    }
  }
];

describe("<Pagination />", () => {
  it("renders with only 1 item", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksWithLength(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain("Loading...");
    await wait();
    wrapper.update();
    const pagination = wrapper.find('div[data-test="pagination"]');
    expect(pagination.text()).toContain("Page 1 of 1");
    expect(toJSON(pagination)).toMatchSnapshot();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(true);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(true);
  });

  it("renders with 5 pages with 18 items", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksWithLength(18)}>
        <Pagination page={1} perPage={4} />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain("Loading...");
    await wait();
    wrapper.update();
    const pagination = wrapper.find('div[data-test="pagination"]');
    expect(pagination.text()).toContain("Page 1 of 5");
    expect(toJSON(pagination)).toMatchSnapshot();
  });

  it("disables prev button when rendering the first page", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksWithLength(18)}>
        <Pagination page={1} perPage={4} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(true);
  });

  it("disables next button when rendering the first page", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksWithLength(18)}>
        <Pagination page={5} perPage={4} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(true);
  });

  it("enables next & prev button when in the middle of pages", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksWithLength(18)}>
        <Pagination page={2} perPage={4} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(false);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(false);
  });
});
