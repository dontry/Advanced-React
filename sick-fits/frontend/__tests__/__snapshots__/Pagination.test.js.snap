// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders with 5 pages with 18 items 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="items?page=0"
      onClick={[Function]}
    >
      ← Prev
    </a>
  </Link>
  <p>
    Page 
    1
     of 
    <span
      className="totalPages"
    >
      5
    </span>
  </p>
  <p>
    18
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      className="next"
      href="items?page=2"
      onClick={[Function]}
    >
      Next →
    </a>
  </Link>
</div>
`;

exports[`<Pagination /> renders with only 1 item 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="items?page=0"
      onClick={[Function]}
    >
      ← Prev
    </a>
  </Link>
  <p>
    Page 
    1
     of 
    <span
      className="totalPages"
    >
      1
    </span>
  </p>
  <p>
    1
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="next"
      href="items?page=2"
      onClick={[Function]}
    >
      Next →
    </a>
  </Link>
</div>
`;
